include/sympack/symPACKMatrix2D.hpp

```diff
@@ -1005,26 +1228,105 @@ namespace symPACK{
+                } else if (d_buf==nullptr) { 
+                    blas::Syrk('U','T',tgt_width, src_snode_size, T(-1.0), pivot_nzval, src_snode_size, beta, buf, ldbuf);
+                } else {
+                    upcxx::when_all(
+                        upcxx::copy(pivot_nzval, d_pivot_nzval, tgt_width*src_snode_size),
+                        upcxx::copy(buf, d_buf, tgt_width * ldbuf)
+                    ).wait();
+
+                    cublas::cublas_syrk_wrapper2(CUBLAS_FILL_MODE_UPPER,CUBLAS_OP_T,tgt_width, src_snode_size,
+                            T(-1.0), symPACK::gpu_allocator.local(d_pivot_nzval), 
+                            src_snode_size, beta, 
+                            symPACK::gpu_allocator.local(d_buf), ldbuf);
+                    CUDA_ERROR_CHECK(cudaDeviceSynchronize());
+
+                    upcxx::copy(d_buf, buf, tgt_width * ldbuf).wait();
+
+                    symPACK::gpu_allocator.deallocate(d_pivot_nzval);
+                    symPACK::gpu_allocator.deallocate(d_buf);
+                }
+
+              } else {
+                    blas::Syrk('U','T',tgt_width, src_snode_size,
+                      T(-1.0), pivot_nzval, src_snode_size, beta, buf, ldbuf);
+              }
+#else
               blas::Syrk('U','T',tgt_width, src_snode_size,
                   T(-1.0), pivot_nzval, src_snode_size, beta, buf, ldbuf);
+#endif                  
               SYMPACK_TIMER_SPECIAL_STOP(UPDATE_SNODE_SYRK);
             }
             else {
               //everything is in row-major
               SYMPACK_TIMER_SPECIAL_START(UPDATE_SNODE_GEMM);
+#ifdef CUDA_MODE
+              if (tgt_width * src_nrows > symPACK::gemm_limit ||
+                  tgt_width * src_snode_size > symPACK::gemm_limit ||
+                  src_nrows * src_snode_size > symPACK::gemm_limit) {
+                dev_ptr d_pivot_nzval = symPACK::gpu_allocator.allocate<T>(tgt_width * src_snode_size);
+                dev_ptr d_facing_nzval = symPACK::gpu_allocator.allocate<T>(src_nrows * src_snode_size);
+                dev_ptr d_buf = symPACK::gpu_allocator.allocate<T>(ldbuf * src_nrows);
+                if (d_pivot_nzval==nullptr) {
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
+                } else if (d_facing_nzval==nullptr) {
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
+                } else if (d_buf==nullptr) { 
+                    blas::Gemm('T','N',tgt_width, src_nrows,src_snode_size,
+                                T(-1.0),pivot_nzval,src_snode_size,
+                                facing_nzval,src_snode_size,beta,buf,ldbuf);
```

Comment:
More leaks here.

Note that `device_allocator::deallocate(nullptr)` is defined as a no-op, so the most elegant solution in this particular case is probably to ensure all three fallback paths (which all seem to take the same action?) all reach the three deallocation calls below (ie by moving them past the enclosing `}`) and let the allocator sort it out.